compiler/z/codegen/SystemLinkagezOS.cpp

```diff
@@ -551,40 +551,61 @@ TR::S390zOSSystemLinkage::generateInstructionsForCall(TR::Node * callNode, TR::R
 
-    // Find GPR6 (xplink) or GPR15 (os linkage) in post conditions
-    TR::Register * systemEntryPointRegister = deps->searchPostConditionRegister(getEntryPointRegister());
-    // Find GPR7 (xplik) or GPR14 (os linkage) in post conditions
-    TR::Register * systemReturnAddressRegister = deps->searchPostConditionRegister(getReturnAddressRegister());
-
-    TR::RegisterDependencyConditions * preDeps = new (trHeapMemory()) TR::RegisterDependencyConditions(deps->getPreConditions(), NULL, deps->getAddCursorForPre(), 0, cg());
+    TR::RegisterDependencyConditions* preDeps = new (trHeapMemory()) TR::RegisterDependencyConditions(deps->getPreConditions(), NULL, deps->getAddCursorForPre(), 0, cg());
+    TR::RegisterDependencyConditions* postDeps = new (trHeapMemory()) TR::RegisterDependencyConditions(NULL, deps->getPostConditions(), 0, deps->getAddCursorForPost(), cg());
 
     if (callNode->getOpCode().isIndirect())
        {
-       generateRRInstruction(cg(), InstOpCode::BASR, callNode, systemReturnAddressRegister, systemEntryPointRegister, preDeps);
+       TR::Register* targetAddress = cg()->evaluate(callNode->getFirstChild());
+
+       generateRSInstruction(cg(), TR::InstOpCode::getLoadMultipleOpCode(), callNode, aeReg, epReg, generateS390MemoryReference(targetAddress, 0, cg()));
+       generateRRInstruction(cg(), InstOpCode::BASR, callNode, raReg, epReg, preDeps);
        callType = TR_XPLinkCallType_BASR;
        }
     else
        {
+       TR::SymbolReference* callSymRef = callNode->getSymbolReference();
+       TR::Symbol* callSymbol = callSymRef->getSymbol();
+
+       if (comp()->isRecursiveMethodTarget(callSymbol))
+          {
+          // No need to load the environment or the entry point for recursive calls because these values are not used
+          // within OMR compiled methods
+          TR::Instruction* callInstr = new (cg()->trHeapMemory()) TR::S390RILInstruction(TR::InstOpCode::BRASL, callNode, raReg, callSymbol, callSymRef, cg());
+          callInstr->setDependencyConditions(preDeps);
+          }
+       else
+          {
+          struct FunctionDescriptor
+             {
+             void* environment;
+             void* func;
+             };
+
+          FunctionDescriptor* fd = reinterpret_cast<FunctionDescriptor*>(callSymbol->castToMethodSymbol()->getMethodAddress());
 
-       TR::SymbolReference *callSymRef = callNode->getSymbolReference();
-       TR::Symbol *callSymbol = callSymRef->getSymbol();
+          genLoadAddressConstant(cg(), callNode, reinterpret_cast<uintptrj_t>(fd), epReg);
+          generateRSInstruction(cg(), TR::InstOpCode::getLoadMultipleOpCode(), callNode, aeReg, epReg, generateS390MemoryReference(epReg, 0, cg()));
+
+          TR::Instruction* callInstr = new (cg()->trHeapMemory()) TR::S390RILInstruction(TR::InstOpCode::BRASL, callNode, raReg, fd->func, callSymRef, cg());
+          callInstr->setDependencyConditions(preDeps);
```

Comment:
Loads the function's address from the function descriptor and dispatches the call.